src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs

```diff
@@ -15,32 +16,69 @@ namespace System.Net.Http.Headers
     // Use HeaderDescriptor.TryGet to resolve an arbitrary header name to a HeaderDescriptor.
     internal readonly struct HeaderDescriptor : IEquatable<HeaderDescriptor>
     {
-        private readonly string _headerName;
-        private readonly KnownHeader? _knownHeader;
+        /// <summary>
+        /// Either a <see cref="KnownHeader"/> or <see cref="string"/>
+        /// </summary>
+        private readonly object? _descriptor;
 
         public HeaderDescriptor(KnownHeader knownHeader)
         {
-            _knownHeader = knownHeader;
-            _headerName = knownHeader.Name;
+            _descriptor = knownHeader;
         }
 
         // This should not be used directly; use static TryGet below
         internal HeaderDescriptor(string headerName)
         {
-            _headerName = headerName;
-            _knownHeader = null;
+            _descriptor = headerName;
         }
 
-        public string Name => _headerName;
-        public HttpHeaderParser? Parser => _knownHeader?.Parser;
-        public HttpHeaderType HeaderType => _knownHeader == null ? HttpHeaderType.Custom : _knownHeader.HeaderType;
-        public KnownHeader? KnownHeader => _knownHeader;
+        public string Name
+        {
+            get
+            {
+                Debug.Assert(_descriptor is not null);
+                object descriptor = _descriptor;
+                return descriptor is KnownHeader knownHeader ?
+                    knownHeader.Name :
+                    Unsafe.As<string>(descriptor);
+            }
+        }
+
+        public object? Descriptor => _descriptor;
```

Comment:
Why is this necessary? Seems dangerous at best...

Reply:
An alternative here might be to have something like 

```csharp
    public ValueTuple<KnownHeader?, string?> KnownHeaderOrName => ...
```

That seems safer because it encapsulates the storage details of this struct.

Reply:
That seems reasonable to avoid some `Unsafe` (at least outside of `HeaderDescriptor` itself). I wonder if it'd have a noticeable perf impact.

Some usages could also be replaced by something like
```c#
public bool IsDefault/IsNotSet => _descriptor is null;

public bool IsKnownHeader(KnownHeader other) => ReferenceEquals(_descriptor, other);
```

Reply:
I removed `Descriptor` and added
```c#
bool HasValue;
bool IsKnownHeader([NotNullWhen(true)] out KnownHeader? knownHeader, [NotNullWhen(false)] out string? headerName);
bool Equals(string other);
bool Equals(KnownHeader other);
```

Thoughts?